traitsui/testing/tester/_dynamic_target_registry.py

```diff
@@ -0,0 +1,85 @@
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Implement the interface of AbstractTargetRegistry to support
+dynamic checking on the target.
+"""
+import inspect
+
+from traitsui.testing.tester._abstract_target_registry import (
+    AbstractTargetRegistry,
+)
+from traitsui.testing.tester.exceptions import (
+    InteractionNotSupported,
+    LocationNotSupported,
+)
+
+
+class DynamicTargetRegistry(AbstractTargetRegistry):
+    """ A registry to support targets with dynamic checks.
+
+    Parameters
+    ----------
+    can_support : callable(target) -> bool
+        A callable that return true if the given target is supported
+        by the registry.
+    interaction_to_handler : dict(type, callable)
+        A dictionary mapping from interaction type to handler callables
+    """
+
+    def __init__(self, *, can_support, interaction_to_handler):
+        self.can_support = can_support
+        self.interaction_to_handler = interaction_to_handler
+
+    def _get_handler(self, target, interaction):
+        """ Reimplemented AbstractTargetRegistry._get_handler """
+        if interaction.__class__ not in self._get_interactions(target):
+            raise InteractionNotSupported(
+                target_class=target.__class__,
+                interaction_class=interaction.__class__,
+                supported=list(self._get_interactions(target)),
+            )
+        return self.interaction_to_handler[interaction.__class__]
+
+    def _get_interactions(self, target):
+        """ Reimplemented AbstractTargetRegistry._get_interactions """
+        if self.can_support(target):
+            return set(self.interaction_to_handler)
+        return set()
+
+    def _get_interaction_doc(self, target, interaction_class):
+        """ Reimplemented AbstractTargetRegistry._get_interaction_doc """
+        if interaction_class not in self._get_interactions(target):
+            raise InteractionNotSupported(
+                target_class=target.__class__,
+                interaction_class=interaction_class,
+                supported=list(self._get_interactions(target)),
+            )
+        return inspect.getdoc(interaction_class)
+
+    def _get_solver(self, target, location):
+        """ Reimplemented AbstractTargetRegistry._get_solver """
+        raise LocationNotSupported(
+            target_class=target.__class__,
+            locator_class=location.__class__,
+            supported=[],
+        )
+
+    def _get_locations(self, target):
+        """ Reimplemented AbstractTargetRegistry._get_locations """
+        return set()
+
+    def _get_location_doc(self, target, locator_class):
+        """ Reimplemented AbstractTargetRegistry._get_location_doc """
+        raise LocationNotSupported(
+            target_class=target.__class__,
+            locator_class=locator_class,
+            supported=[],
+        )
```

Comment:
Not for this PR, but will we want to add a section to the documentation about `DynamicTargetRegsitry`? This is in a privately named module so I assume you were thinking not / we don't expect users to mess with it.

As far as users wanting to extend UITester functionality, I feel like I could imagine scenarios where it would be useful to work with an additional `DynamicTargetRegistry` rather than a `TargetRegistry` to append to the default registries already provided on a UITester.  For example right now, if this went in and say hypothetically we never put in an `IsVisible` implementation along with this.  A downstream user may want to do something like that? 
However, I can also see an argument for this only being used internally, and keeping the means of extending UITester functionality exactly as it is now.

Reply:
I think I started out keeping this internal, but I am totally open to exposing this publicly if it is useful downstream. With this made public, then we need to think harder about the name...